Link directly to files stored in the Media Library using the Link UI

getdave opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This PR includes media files (attachments) in the results returned by `__experimentalFetchLinkSuggestions`. This means that you can search for and see results for media files in the Link UI.

This means you can now link _directly_ to files that have been uploaded to the WordPress Media library.


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Many folks want to be able to link to a Media file directly from the Link UI. Currently [this is not possible](https://github.com/WordPress/gutenberg/issues/8322) without significant workarounds.

A key use case is the ability to link to previously uploaded PDFs which is a very common requirements.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This PR appends an additional query which makes a request to the [Media REST endpoint](https://developer.wordpress.org/rest-api/reference/media/) using the `search` query parameter to find attachments which match the given query.

Given that the Search and Media endpoints are very similar it's then simple enough to normalize the results into the format expected to be returned by `__experimentalFetchLinkSuggestions`.

The result of this is that you can now search for media files directly from the Link UI.

Please note: this PR [doesn't provide a means to upload media directly from the Link UI itself](https://github.com/WordPress/gutenberg/pull/37260) (that's something for another day).

## Testing Instructions

- Upload some media files to your WP installation.
- Give the uploaded media files titles (via the Media Library fields - _not_ the filenames) that you can remember in order that you can search for them later.
- Add a paragraph block and add some text
- Create a link and search using a query that matches one of the media files you previously uploaded.
- See a result matching your media file
- Check various types of Media that are valid in WordPress.

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/444434/159754434-12d4e439-3a26-428a-9efb-23fdf2853a15.mov

getdave committed:
Add Media endpoint

draganescu commented on packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js:
```diff
@@ -153,6 +153,27 @@ const fetchLinkSuggestions = async (
 		);
 	}
 
+	if ( ! type || type === 'media' ) {
```
Why do we use `! type` everywhere here?

getdave replied:
If you _don't_ provide a type then it should include all queries.

getdave replied:
So it's a filter. If you don't pass `type` then it will query all the available types. However, if you pass a `type` then it will filter by a given type only (e.g. `post`, `taxonomy`...etc). See tests for examples.

draganescu commented:
This is so cool @getdave. This PR looks very good. Anything in particular that makes it a draft?

getdave commented:
> This is so cool @getdave. This PR looks very good. Anything in particular that makes it a draft?

Not really. I just knocked it up in 5mins last night so I wasn't confident in saying it's ready for review.

getdave commented:
This PR is not currently my main focus but I will make my best effort to advance it as I see it's important to users.

getdave committed:
Update test to account for attachment media

getdave commented:
cc'ing @mintplugins as the original creator of [the Issue](https://github.com/WordPress/gutenberg/issues/8322). I appreciate it was a long time ago but if you have time/inclination you can [follow these steps to test this PR](https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/).

paaljoachim commented:
Thank you very much, Dave!

This is very useful!
Being able to write the name of a media file in the link search box is very helpful!
A very natural feature to have included!

It works really well!

![Screenshot 2022-03-25 at 07 25 21](https://user-images.githubusercontent.com/5323259/160066396-a2d1a2cc-af03-405d-b5e4-05154ad22934.jpg)

paaljoachim reviewed (commented):
Feature works well for me. Thank you!

jasmussen commented:
Oh sweet! I guess this one just needs a code review to land.

As with every piece of UI in the entire interface, there's always an opportunity to improve the visuals. The following should not block this PR at all, but just sharing in terms of future musings:

<img width="1220" alt="Searching" src="https://user-images.githubusercontent.com/1204802/160076218-b727f2f2-d757-4280-9469-c00cbd0f5461.png">

(For that design, we could probably use a general File icon, but also a note to self, might be nice with a bespoke PDF variant as well.)

paaljoachim commented:
Hey @jasmussen Joen.

I will just add it in...
Seeing the visual using icons without the page/post/attachment text afterward creates for me a hesitation of what the icon actually means. Having the text clearly tells me what is what + the link also tells me the file url which is very helpful. 
The new design lacks clarity as to what is what and can easily create confusion. We could perhaps hover over each object to get the "old" information back. But this is a discussion for another time.

jasmussen commented:
It would be a future exploration, and I think figuring out what works will feel much more natural in a PR.

scruffian commented on packages/core-data/src/fetch/test/__experimental-fetch-link-suggestions.js:
```diff
@@ -70,6 +70,18 @@ jest.mock( '@wordpress/api-fetch', () =>
+					{
+						id: 54,
+						title: {
+							rendered: 'Media',
```
Might be worth making the title less generic, just in case a bug makes the API return the word "Media" for all items?

scruffian replied:
Or maybe that's what it does?!

getdave replied:
I think you made a good point. I made the title more generic.

getdave replied:
Fixed.

scruffian commented on packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js:
```diff
@@ -153,6 +153,27 @@ const fetchLinkSuggestions = async (
 		);
 	}
 
+	if ( ! type || type === 'attachment' ) {
```
why `! type`?

scruffian replied:
Oh, is type a filter? Makes sense now...

scruffian commented on packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js:
```diff
@@ -153,6 +153,27 @@ const fetchLinkSuggestions = async (
+						};
+					} );
+				} )
+				.catch( () => [] )
```
Should we throw an error here?

getdave replied:
Probably needs looking into. The pattern is already in evidence so I'm following that for now.

scruffian reviewed (approved):
This looks great, just a few nitpicks but nothing to stop you merging.

getdave committed:
Improve commenting of catch statement

getdave committed:
Make Media title less genric

Addresses https://github.com/WordPress/gutenberg/pull/39701#discussion_r835044126

getdave commented:
@paaljoachim @jasmussen Thank you both for taking time to review this and make such good suggestions 🙇 

The icon idea is certainly worth exploring in a followup but I also agree with Paal that we ought to retain some kind of text-based title.
